dev/src/transaction.ts

```diff
@@ -449,57 +472,46 @@ export class Transaction implements firestore.Transaction {
     documentRef: DocumentReference<any, any>,
     precondition?: firestore.Precondition
   ): this {
-    if (this._readTime) {
+    if (!this._writeBatch) {
       throw new Error(READ_ONLY_WRITE_ERROR_MSG);
     }
     this._writeBatch.delete(documentRef, precondition);
     return this;
   }
 
-  /**
-   * Starts a transaction and obtains the transaction id from the server.
-   *
-   * @private
-   * @internal
-   */
-  async begin(): Promise<void> {
-    const request: api.IBeginTransactionRequest = {
-      database: this._firestore.formattedName,
-    };
-
-    if (this._readOnly) {
-      request.options = {
-        readOnly: {},
-      };
-    } else if (this._transactionId) {
-      request.options = {
-        readWrite: {
-          retryTransaction: this._transactionId,
-        },
-      };
-    }
-
-    const resp = await this._firestore.request<
-      api.IBeginTransactionRequest,
-      api.IBeginTransactionResponse
-    >('beginTransaction', request, this._requestTag);
-    this._transactionId = resp.transaction!;
-  }
-
   /**
    * Commits all queued-up changes in this transaction and releases all locks.
    *
    * @private
    * @internal
    */
   async commit(): Promise<void> {
-    if (this._readTime) {
+    if (!this._writeBatch) {
       throw new Error(READ_ONLY_WRITE_ERROR_MSG);
     }
-    await this._writeBatch._commit({
-      transactionId: this._transactionId,
-      requestTag: this._requestTag,
-    });
+
+    // If we have not performed any reads in this particular attempt
+    // then the write will be atomically committed without a transaction
+    let transactionId: Uint8Array | undefined;
+    if (this._transactionIdPromise) {
+      transactionId = await this._transactionIdPromise;
+    }
+
+    try {
+      await this._writeBatch._commit({
+        transactionId,
+        requestTag: this._requestTag,
+      });
+      this._transactionIdPromise = undefined;
+      this._prevTransactionId = transactionId;
+    } catch (err) {
+      // A failed commit also rolls back the transaction so if we know that the
+      // server responded with an error then we clear the transaction ID so that
+      // we don't continue on to perform a superfluous rollback
+      this._transactionIdPromise = undefined;
+      this._prevTransactionId = transactionId;
+      throw err;
```

Comment:
I feel that this statement is true just from my experimentation with the API, but it is not documented... is it correct? I need to determine if the error was an "error response" from the server and not some other comms error for example, what is the best way to determine this?